File: orchestrator/session.py
import subprocess
import shutil
from dataclasses import dataclass
from orchestrator import log

logger = log.get("session")


@dataclass
class SessionResult:
    returncode: int
    stdout: str
    stderr: str


def _build_cmd(prompt: str) -> list[str]:
    return ["opencode", prompt]


def run(prompt: str, timeout: int | None = None) -> SessionResult:
    cmd = _build_cmd(prompt)
    logger.info("running opencode with %d-byte prompt", len(prompt))

    if not shutil.which("opencode"):
        raise FileNotFoundError(
            "opencode executable not found on $PATH. "
            "Install it via: pip install opencode"
        )

    proc = subprocess.Popen(
        cmd,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )
    try:
        stdout, stderr = proc.communicate(timeout=timeout)
    except subprocess.TimeoutExpired:
        proc.kill()
        stdout, stderr = proc.communicate()
        logger.warning("opencode timed out after %ds (code %d)", timeout or 0, proc.returncode)
        return SessionResult(returncode=-9, stdout=stdout, stderr=stderr)

    logger.info("opencode exited with code %d", proc.returncode)
    return SessionResult(
        returncode=proc.returncode,
        stdout=stdout,
        stderr=stderr,
    )

File: orc/internal/session/session.go
package session

import (
	"fmt"
	"io"
	"os"
	"os/exec"
	"strings"
	"time"
)

var execCommand = exec.Command

type Result struct {
	ReturnCode int
	Stdout     string
	Stderr     string
}

func Run(prompt string, timeout time.Duration) (*Result, error) {
	path, err := exec.LookPath("opencode")
	if err != nil {
		return nil, fmt.Errorf("opencode not found on $PATH: install via 'pip install opencode' or download from https://opencode.ai")
	}
	fmt.Fprintf(os.Stderr, "command: opencode run --dangerously-skip-permissions (prompt=%d bytes)\n", len(prompt))

	cmd := execCommand(path, "run", "--dangerously-skip-permissions", prompt)
	cmd.Stdin = os.Stdin

	var stdout, stderr strings.Builder
	cmd.Stdout = io.MultiWriter(&stdout, os.Stdout)
	cmd.Stderr = io.MultiWriter(&stderr, os.Stderr)

	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("starting opencode: %w", err)
	}

	done := make(chan error, 1)
	go func() {
		done <- cmd.Wait()
	}()

	if timeout > 0 {
		select {
		case <-done:
		case <-time.After(timeout):
			cmd.Process.Kill()
			<-done
			return &Result{
				ReturnCode: -9,
				Stdout:     stdout.String(),
				Stderr:     stderr.String(),
			}, nil
		}
	} else {
		<-done
	}

	return &Result{
		ReturnCode: cmd.ProcessState.ExitCode(),
		Stdout:     stdout.String(),
		Stderr:     stderr.String(),
	}, nil
}
